src/python/platforms/fuchsia/undercoat.py

```diff
@@ -0,0 +1,153 @@
+        'Error running undercoat command %s: %s' % (args, result.output))
+
+  return result
+
+
+def undercoat_instance_command(command, handle, *args):
+  """Helper for the subset of undercoat commands that operate on an instance."""
+  try:
+    return undercoat_api_command(command, '-handle', handle, *args)
+  except UndercoatError:
+    # Try to print extra logs and shut down
+    # TODO(eep): Should we be attempting to automatically restart?
+    dump_instance_logs(handle)
+    stop_instance(handle)
+    raise
+
+
+def dump_instance_logs(handle):
+  """Dump logs from an undercoat instance."""
+  # Avoids using undercoat_instance_command in order to avoid recursion on error
+  qemu_log = undercoat_api_command('get_logs', '-handle', handle).output
+  logs.log_warn(qemu_log)
+
+
+def start_instance():
+  """Start an instance via undercoat."""
+  handle = undercoat_api_command('start_instance').output.strip()
+  logs.log('Started undercoat instance with handle %s' % handle)
+
+  # Immediately save the handle in case we crash before stop_instance()
+  # is called
+  shell.create_directory(UNDERCOAT_HANDLE_DIR)
+  with tempfile.NamedTemporaryFile(
+      dir=UNDERCOAT_HANDLE_DIR, mode='w', delete=False) as f:
+    f.write(handle)
+
+  return handle
+
+
+def stop_all():
+  """Attempt to stop any running undercoat instances that may have not been
+  cleanly shut down."""
+  for handle_file, handle in get_all_handles():
+    try:
+      undercoat_instance_command('stop_instance', handle)
+    except UndercoatError:
+      pass
+
+    # Even if we failed to stop_instance above, there's no point in trying
+    # again later
+    shell.remove_file(handle_file)
+
+
+def stop_instance(handle):
+  """Stop a running undercoat instance."""
+  # Avoids using undercoat_instance_command in order to avoid recursion on error
+  result = undercoat_api_command('stop_instance', '-handle', handle)
+
+  # Remove the corresponding handle file now that we've cleanly shut down
+  handle_file = next((f for f, h in get_all_handles() if h == handle), None)
+  if handle_file:
+    shell.remove_file(handle_file)
+
+  return result
+
+
+def list_fuzzers(handle):
+  """Start an instance via undercoat."""
+  return undercoat_instance_command('list_fuzzers', handle).output.split('\n')
+
+
+def prepare_fuzzer(handle, fuzzer):
+  """Prepare a fuzzer of the given name for use, via undercoat."""
+  return undercoat_instance_command('prepare_fuzzer', handle, '-fuzzer', fuzzer)
+
+
+def run_fuzzer(handle, fuzzer, outdir, args):
+  """Run a fuzzer of the given name, via undercoat."""
+  # TODO(fxbug.dev/47490): Pass back raw return code from libFuzzer?
+  undercoat_args = ['-fuzzer', fuzzer]
+  if outdir:
+    undercoat_args += ['-artifact-dir', outdir]
+  args = undercoat_args + ['--'] + args
+  return undercoat_instance_command('run_fuzzer', handle, *args)
+
+
+def put_data(handle, fuzzer, src, dst):
+  """Put files for a fuzzer onto an instance, via undercoat."""
```

Comment:
Can you add some documentation on the semantics here when `src` or `dst` is a directory (or both)?